core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -460,9 +460,9 @@ public void updateShardState(final ShardRouting newRouting,
                      * incremented.
                      */
                     // to prevent primary relocation handoff while resync is not completed
-                    boolean resyncStarted = primaryReplicaResyncInProgress.compareAndSet(false, true);
-                    if (resyncStarted == false) {
-                        throw new IllegalStateException("cannot start resync while it's already in progress");
+                    final IndexShardState prevState = changeState(IndexShardState.PROMOTING, "Promoting to primary");
```

Comment:
instead of changing the state first and then checking whether the previous state was the right one, let's only change the state if the current state matches (note that we're under the mutex here already, so it's safe to do this).